File: web/src/switch.ts
import { existsSync } from "node:fs";
import { readFile, readdir, writeFile } from "node:fs/promises";
import { join } from "node:path";

import { checkoutTree } from "./pull.ts";
import { setUpstreamBranch } from "./push.ts";
import { updateBranch, type PackfileCache } from "./utils.ts";

export async function switchBranch(path: string, branchName: string): Promise<void> {
	const gitDir = join(path, ".git");

	if (!existsSync(gitDir)) {
		throw new Error("Not a git repository");
	}

	const cache: PackfileCache = new Map();

	// 1. Local branch exists: check out its tree, then point HEAD at it.
	const localBranchPath = join(gitDir, "refs", "heads", branchName);
	if (existsSync(localBranchPath)) {
		const commitSha = (await readFile(localBranchPath, "utf-8")).trim();

		// Check out the tree first (uses the current HEAD as the baseline for
		// change detection); only move HEAD once the checkout succeeds.
		await checkoutTree(gitDir, path, commitSha, cache);

		const headPath = join(gitDir, "HEAD");
		await writeFile(headPath, `ref: refs/heads/${branchName}\n`, "utf-8");
		return;
	}

	// 2. DWIM: no local branch, but exactly one remote tracking branch exists.
	const dwim = await findRemoteBranch(gitDir, branchName);
	if (dwim) {
		const { remoteName, commitSha } = dwim;

		await updateBranch(gitDir, branchName, commitSha);
		await setUpstreamBranch(path, remoteName, branchName);

		await checkoutTree(gitDir, path, commitSha, cache);

		const headPath = join(gitDir, "HEAD");
		await writeFile(headPath, `ref: refs/heads/${branchName}\n`, "utf-8");
		return;
	}

	// 3. No local branch and zero (or multiple) matching remotes.
	throw new Error(`Branch '${branchName}' not found`);
}

/**
 * Finds a single remote that has `refs/remotes/<remote>/<branchName>`.
 * Returns `{ remoteName, commitSha }` when exactly one match exists, otherwise null.
 */
async function findRemoteBranch(
	gitDir: string,
	branchName: string,
): Promise<{ remoteName: string; commitSha: string } | null> {
	const remotesDir = join(gitDir, "refs", "remotes");
	if (!existsSync(remotesDir)) {
		return null;
	}

	const remoteNames = await readdir(remotesDir);
	let match: { remoteName: string; commitSha: string } | null = null;
	let matchCount = 0;

	for (const remoteName of remoteNames) {
		const branchRef = join(remotesDir, remoteName, branchName);
		if (!existsSync(branchRef)) {
			continue;
		}
		const sha = (await readFile(branchRef, "utf-8")).trim();
		match = { remoteName, commitSha: sha };
		matchCount += 1;
	}

	return matchCount === 1 ? match : null;
}

File: dotnet/Gitologist/Switch.cs
namespace Gitologist;

public static class Switch
{
    public static async Task SwitchBranch(string path, string branchName)
    {
        var gitDir = Path.Combine(path, ".git");

        if (!Directory.Exists(gitDir))
        {
            throw new InvalidOperationException("Not a git repository");
        }

        var cache = new Utils.PackfileCache();

        // 1. Local branch exists: check out its tree, then point HEAD at it.
        var localBranchPath = Path.Combine(gitDir, "refs", "heads", branchName);
        if (File.Exists(localBranchPath))
        {
            var commitSha = (await File.ReadAllTextAsync(localBranchPath)).Trim();

            // Check out the tree first (uses the current HEAD as the baseline for
            // change detection); only move HEAD once the checkout succeeds.
            await Pull.CheckoutTree(gitDir, path, commitSha, cache);

            var headPath = Path.Combine(gitDir, "HEAD");
            await File.WriteAllTextAsync(headPath, $"ref: refs/heads/{branchName}\n");
            return;
        }

        // 2. DWIM: no local branch, but exactly one remote tracking branch exists.
        var dwim = FindRemoteBranch(gitDir, branchName);
        if (dwim != null)
        {
            var (remoteName, commitSha) = dwim.Value;

            await Utils.UpdateBranch(gitDir, branchName, commitSha);
            await Push.SetUpstreamBranch(path, remoteName, branchName);

            await Pull.CheckoutTree(gitDir, path, commitSha, cache);

            var headPath = Path.Combine(gitDir, "HEAD");
            await File.WriteAllTextAsync(headPath, $"ref: refs/heads/{branchName}\n");
            return;
        }

        // 3. No local branch and zero (or multiple) matching remotes.
        throw new InvalidOperationException($"Branch '{branchName}' not found");
    }

    /// <summary>
    /// Finds a single remote that has <c>refs/remotes/&lt;remote&gt;/&lt;branchName&gt;</c>.
    /// Returns <c>(remoteName, commitSha)</c> when exactly one match exists, otherwise null.
    /// </summary>
    private static (string remoteName, string commitSha)? FindRemoteBranch(string gitDir, string branchName)
    {
        var remotesDir = Path.Combine(gitDir, "refs", "remotes");
        if (!Directory.Exists(remotesDir))
        {
            return null;
        }

        (string remoteName, string commitSha)? match = null;
        var matchCount = 0;

        foreach (var remoteDir in Directory.GetDirectories(remotesDir))
        {
            var branchRef = Path.Combine(remoteDir, branchName);
            if (!File.Exists(branchRef))
            {
                continue;
            }
            var sha = File.ReadAllText(branchRef).Trim();
            match = (Path.GetFileName(remoteDir), sha);
            matchCount++;
        }

        return matchCount == 1 ? match : null;
    }
}
